### PR TITLE
chore(cli): add cli/deploy pkg

### PR DESCRIPTION
**Pull request by iamhopaul123**
By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the Apache 2.0 License.

### CODE REVIEW

File: internal/pkg/cli/deploy/deploy.go
Comment (efekarakus) [inline internal/pkg/cli/deploy/deploy.go]: Instead of having the empty interface and accepting any manifest types I wonder if we should have specific structs instead?

```go
type RDWSDeployer struct { ...}
type LBWSDeployer struct { ... }
```

File: internal/pkg/cli/deploy/deploy.go
Comment (efekarakus) [inline internal/pkg/cli/deploy/deploy.go:155]: should the image tag be part of the `DeployWorkloadInput` instead?

File: internal/pkg/cli/deploy/deploy.go
Comment (efekarakus) [inline internal/pkg/cli/deploy/deploy.go]: To be consistent with the aws sdk go, can we use singular? (example: https://docs.aws.amazon.com/sdk-for-go/api/service/ecs/#CreateCapacityProviderInput)

```suggestion
type UploadArtifactsInput struct {
```

File: internal/pkg/cli/deploy/deploy.go
Comment (efekarakus) [inline internal/pkg/cli/deploy/deploy.go]: nit:
```suggestion
	FS                 *afero.Afero
```

File: internal/pkg/cli/deploy/deploy.go
Comment (efekarakus) [inline internal/pkg/cli/deploy/deploy.go]:
```suggestion
	WW                 WorkspaceReader
```

File: internal/pkg/cli/deploy/deploy.go
Comment (efekarakus) [inline internal/pkg/cli/deploy/deploy.go]: What does this mean? I guess my confusion is around why have both `ImageDigest` and `BuildRequired`

File: internal/pkg/cli/deploy/deploy.go
Comment (efekarakus) [inline internal/pkg/cli/deploy/deploy.go]: Do all workloads need this much info? (It's okay if it does!)

Or is there an opportunity to do something like:
```go
type DeployRDWSInput struct { ...}
func (in *DeployRDWSInput) Deploy(deployer *WorkloadDeployer) error

type DeployLBWSInput struct { ...}
func (in *DeployLBWSInput) Deploy(deployer *WorkloadDeployer) error
```

File: internal/pkg/cli/deploy/deploy.go
Comment (efekarakus) [inline internal/pkg/cli/deploy/deploy.go:273]: should we have a `logger` interface or something part of `WorkloadDeployer`?

**Review by efekarakus — COMMENTED** [review]
I haven't reviewed completely, just leaving some initial feedback to hear your thoughts

File: internal/pkg/cli/deploy/deploy.go
Comment (iamhopaul123) [reply internal/pkg/cli/deploy/deploy.go]: `mft` is just a private cache variable not an input and the deployer reads from the workspace. If we have different deployer type then we'll have to handle the manifest reading in each of our subcommand. Could be something duplicated right 🤔

File: internal/pkg/cli/deploy/deploy.go
Comment (iamhopaul123) [reply internal/pkg/cli/deploy/deploy.go:155]: It is also used when building and uploading the image.

File: internal/pkg/cli/deploy/deploy.go
Comment (iamhopaul123) [reply internal/pkg/cli/deploy/deploy.go]: good idea!

File: internal/pkg/cli/deploy/deploy.go
Comment (iamhopaul123) [reply internal/pkg/cli/deploy/deploy.go]: but afero also uses `Fs` 🤔

File: internal/pkg/cli/deploy/deploy.go
Comment (iamhopaul123) [reply internal/pkg/cli/deploy/deploy.go]: Do you mean `WS`?

File: internal/pkg/cli/deploy/deploy.go
Comment (efekarakus) [reply internal/pkg/cli/deploy/deploy.go]: it's okay, just a nit but i bet afero probably regrets it, see stdlib: https://pkg.go.dev/io/fs#FS

Comment (iamhopaul123) [reply internal/pkg/cli/deploy/deploy.go]: ok gotcha i'll make the change

File: internal/pkg/cli/deploy/deploy.go
Comment (iamhopaul123) [reply internal/pkg/cli/deploy/deploy.go]: because of https://github.com/aws/copilot-cli/blob/mainline/internal/pkg/cli/svc_deploy.go#L409-L410

but maybe we don't need the buildrequired at all as an output

File: internal/pkg/cli/deploy/deploy.go
Comment (iamhopaul123) [reply internal/pkg/cli/deploy/deploy.go:273]: I thought about it but it might be ok since it is within the `cli` package so that we don't output logs everywhere. Why do you want to add a `logger` interface? Like what functionality you think we should have here?

File: internal/pkg/manifest/svc.go
Comment (huanjani) [inline internal/pkg/manifest/svc.go]: tiny nit: 
```suggestion
// IsTypeAService returns if manifest type is service.
```

**Comment by huanjani** [comment]
This is exciting! I remember you mentioning your plan to do this, but would you mind reminding me what the purpose/goal is? Is it to separate out the cli level-specific parts of the existing deploy package? Is this related to the `package` command work?

**Comment by iamhopaul123** [comment]
> This is exciting! I remember you mentioning your plan to do this, but would you mind reminding me what the purpose/goal is? Is it to separate out the cli level-specific parts of the existing deploy package? Is this related to the `package` command work?

Yes. But this change only includes two public methods for the `workload deploy`. I'll add another one for `package` (see design doc for the new implementation plan). However, some private methods and functions will be shared for the last one.

**Review by huanjani — APPROVED** [review]
Nice work! Thanks for doing this.

File: internal/pkg/cli/deploy/deploy.go
Comment (efekarakus) [inline internal/pkg/cli/deploy/deploy.go]: This field is a bit odd, can we remove it? Why call the method if this field is false, the client already has that information so that should be able to do on their own:
```go
if shouldUploadArtifacts {
    d.UploadArtifacts(...)
}
```

File: internal/pkg/cli/deploy/deploy.go
Comment (efekarakus) [inline internal/pkg/cli/deploy/deploy.go]: Do these need to be provided by the client?  They're fields used for testing so pushing the clients to provide the values for these are a bit awkward.

Can we instead have them be private fields in `DeployWorkload` that get set in the constructor, or alternatively:
```suggestion
	ManifestInterpolator     Interpolator
	Manifest                  manifest.WorkloadManifest
```
If you'd like to keep them as inputs to `UploadArtifact`

File: internal/pkg/cli/deploy/deploy.go
Comment (efekarakus) [inline internal/pkg/cli/deploy/deploy.go:213]: Dumb question: what was the difference between these two? 🤔

File: internal/pkg/cli/deploy/deploy.go
Comment (efekarakus) [inline internal/pkg/cli/deploy/deploy.go]: Should we accepting an interface instead for ease of testing? Maybe:
```go
type FileReader interface {
   ReadFile(string) ([]byte, error)
}
```

File: internal/pkg/cli/deploy/deploy.go
Comment (efekarakus) [inline internal/pkg/cli/deploy/deploy.go]: Same feedback as above

File: internal/pkg/cli/deploy/deploy.go
Comment (efekarakus) [inline internal/pkg/cli/deploy/deploy.go]: If these fields are private how will the client set them 🤔 ?  
My preference is for the `cli/deploy` package not to have any `newXXXX` fields to create clients. Instead, we should have dependency injection from the `cli` package.
1. The `cli` package is responsible for creating these objects
2. The `cli/deploy` package accepts only interfaces or data types:   
```suggestion
	CustomResources      CustomResourcesUploader
	S3                   Uploader
```

File: internal/pkg/cli/deploy/deploy.go
Comment (efekarakus) [inline internal/pkg/cli/deploy/deploy.go]: nit: doc comments 🥺

File: internal/pkg/cli/deploy/deploy.go
Comment (efekarakus) [inline internal/pkg/cli/deploy/deploy.go]: nit: should we move this below `ForceNewUpdate` since it's a data type and not an interface

File: internal/pkg/cli/deploy/deploy.go
Comment (efekarakus) [inline internal/pkg/cli/deploy/deploy.go:233]: Why do we have a specific field for `RDWS` for an output that should work for any workload types 🤔 ?   
I feel like this is an opportunity to use polymorphism (kinda like: https://refactoring.com/catalog/replaceConditionalWithPolymorphism.html)

```go
type workloadDeployOutput struct{ ... }
func (out workloadDeployOutput) Subscriptions() []manifest.TopicSubscription

type rdwsDeployOutput struct{ ... }
func (out rdwsDeployOutput) Subscriptions() []manifest.TopicSubscription
func (out rdwsDeployOutput) Endpoint() string

type lbwsDeployOutput struct{ ... }
func (out lbwsDeployOutput) Subscriptions() []manifest.TopicSubscription
func (out lbwsDeployOutput) Endpoint() string
```

This way all the output types implement the same interface `type output interface { Subscriptions() []manifest.TopicSubscription }` and some have even more methods:
```go
type addressableOutput interface {
   output
   Endpoint() string
}
if aout, ok := out.(addressableOutput) {
   log(aout.Endpoint())
}
log(out.Subscriptions())
```

File: internal/pkg/cli/deploy/deploy.go
Comment (efekarakus) [inline internal/pkg/cli/deploy/deploy.go]: Should custom resources be included in this method?

File: internal/pkg/cli/deploy/deploy.go
Comment (efekarakus) [reply internal/pkg/cli/deploy/deploy.go:273]: Sounds good! I didn't actually realize that this pkg was under the `cli` directory 😅

**Review by efekarakus — COMMENTED** [review]
Sorry for the late review boss! Thanks for moving all this logic to its own package 👍 It's a great first step that leaves the codebase better than before. 

I think what I care the most about is: Having the **exported** names (types, methods) be the best they can be (dependency injection, good type names).   
The rest, body of the methods and private stuff, can remain as is.

After the next revision, I can approve!

File: internal/pkg/cli/deploy/deploy.go
Comment (iamhopaul123) [reply internal/pkg/cli/deploy/deploy.go]: agreed! i'll keep them as private vars

File: internal/pkg/cli/deploy/deploy.go
Comment (iamhopaul123) [reply internal/pkg/cli/deploy/deploy.go:213]: `AppTags` are the app level resource tags while `resourceTags` are tags assigned during `workload deploy`. We'll merge these two groups.

File: internal/pkg/cli/deploy/deploy.go
Comment (iamhopaul123) [reply internal/pkg/cli/deploy/deploy.go]: i'll put it into the constructor and make it to be a private field

File: internal/pkg/cli/deploy/deploy.go
Comment (iamhopaul123) [reply internal/pkg/cli/deploy/deploy.go:233]: Hmm I did this because if you look at aws sdk, the output of their APIs leave some fields empty if they are not applicable. For example, if the service is Fargate then some EC2 fields are not set in the output. Also, I thought it would be easier to use as well. What advantage do you think over the current implementation 🤔

File: internal/pkg/cli/deploy/deploy.go
Comment (iamhopaul123) [reply internal/pkg/cli/deploy/deploy.go]: Good callout but not for now. I'll add it once I add the feature allowing `workload package` to be able to generate CFN template/params that are ready to deploy. Before doing that, I'll do the refactoring keeping the current behavior.

File: internal/pkg/cli/deploy/deploy.go
Comment (iamhopaul123) [reply internal/pkg/cli/deploy/deploy.go]: But the problem is `manifest.UnmarshalWorkload` doesn't have any receiver 🤔

File: internal/pkg/cli/deploy/deploy.go
Comment (efekarakus) [inline internal/pkg/cli/deploy/deploy.go]: nit: 
```suggestion
type WorkloadDeployerInput struct {
```

File: internal/pkg/cli/deploy/deploy.go
Comment (efekarakus) [inline internal/pkg/cli/deploy/deploy.go]: Is there a reason why we cant do this instead:
```suggestion
	ForceUpdater          ServiceForceUpdater
	AppVersionGetter      VersionGetter
```

File: internal/pkg/cli/deploy/deploy.go
Comment (efekarakus) [reply internal/pkg/cli/deploy/deploy.go:233]: I'm not strongly opposed to it given that it's just one field at the moment, so if you want to feel free to proceed.  

But I don't agree that it's easier to use 🤔  How will this type progress in the future? My guess is like this:
```go
type DeployWorkloadOutput struct {
	RDWSAlias     string
	LBWSAlias     string
	BackendAlias string
	Subscriptions []manifest.TopicSubscription
}
```
Now each one of these fields need to be individually checked for their zero-value by the client, the `DeployWorkloadOutput` is a mixed bag of mutually exclusive fields. Instead compare it to using duck-typing: https://gist.github.com/efekarakus/6dd25c0dfc11badf5d506dd177bf881d

**Review by efekarakus — APPROVED** [review]
Looks good, approving and adding the label to unblock you!

Two comments below, let me know what you think 🧠

File: internal/pkg/cli/deploy/deploy.go
Comment (iamhopaul123) [reply internal/pkg/cli/deploy/deploy.go:233]: Ohh gotcha. I think it's because i haven't put it into the real production yet. I'll make this change when integrating this new package and removing the old codes.

File: internal/pkg/cli/deploy/deploy.go
Comment (iamhopaul123) [reply internal/pkg/cli/deploy/deploy.go]: For `NewSvcUpdater` because we won't know what updater we should initialize.

File: internal/pkg/cli/deploy/deploy.go
Comment (iamhopaul123) [reply internal/pkg/cli/deploy/deploy.go:233]: Seems like it is not easy to do what you proposed above because the order of recommendation becomes a problem :( let me know your thought on [this PR](https://github.com/aws/copilot-cli/pull/3239)!

File: internal/pkg/cli/deploy/deploy.go
Comment (Lou1415926) [inline internal/pkg/cli/deploy/deploy.go:165]: Sorry for being so late to the PR!!! 🙇🏼‍♀️ Just a concern to throw out here: I wonder if it should be considered a side effect that we are actually reading/applying env override/validating the manifest here, in a constructor 🤔 I feel like constructor, ideally should only init a few members, without doing the extra work mentioned just now.

Comment (iamhopaul123) [reply internal/pkg/cli/deploy/deploy.go:165]: The reason why I put the manifest reading in the constructor is because it is used in both `UploadArtifacts` and `DeployWorkload` (`GenerateCloudFormationTemplate` for `workload package` in future PR as well). If we don't handle the logic in the constructor, then we have to either 1) define it as a function within the `cli` package and call it for `workload deploy/package` or 2) call `workloadManifest` within each of the `workloadDeployer` method. What do you think?

Comment (Lou1415926) [reply internal/pkg/cli/deploy/deploy.go:165]: Ya I like the first option! I feel like it makes sense to not delegate the manifest responsibility to `deploy` package, since manifest operations are not unique to deployment - it is also needed for packaging (is this correct?)

Comment (iamhopaul123) [reply internal/pkg/cli/deploy/deploy.go:165]: > I feel like it makes sense to not delegate the manifest responsibility to deploy package

That makes sense to me. I'll make the change in the new PR.

> since manifest operations are not unique to deployment - it is also needed for packaging (is this correct?)

yes it is required but the thing is i put here also because it is used in `packaging` since as I mentioned `GenerateCloudFormationTemplate` also needs the applied manifest.

File: internal/pkg/cli/deploy/deploy.go
Comment (iamhopaul123) [reply internal/pkg/cli/deploy/deploy.go:233]: Nvm boss updated the new PR to address this fb

**Pull request merged**
